chcemvediet/apps/inforequests/admin.py

```diff
@@ -385,20 +389,69 @@ def get_inforequest(self, obj):
         if obj.type in [Action.TYPES.REQUEST, Action.TYPES.ADVANCED_REQUEST]:
-            dependency.append(format_html(
+            constraints.append(format_html(
                 u'{} is type {}.'.format(admin_obj_format(obj), obj.get_type_display())))
         if len(obj.branch.actions) == 1:
-            dependency.append(format_html(
+            constraints.append(format_html(
                 u'{} is the only action in the branch.'.format(admin_obj_format(obj))))
-        return dependency
+        return constraints
 
     def render_delete_form(self, request, context):
         context[u'delete_constraints'] = self.delete_constraints(context[u'object'])
         context[u'ADMIN_EXTEND_SNOOZE_BY_DAYS'] = ADMIN_EXTEND_SNOOZE_BY_DAYS
         return super(ActionAdmin, self).render_delete_form(request, context)
 
+    @transaction.atomic
+    def delete_selected(self, request, queryset):
+        constraints = []
+        warnings = []
+        outbound = InforequestEmail.objects.none()
+        inbound = InforequestEmail.objects.none()
+        for obj in queryset:
+            obj_constraints = self.delete_constraints(obj)
+            constraints += obj_constraints
+            inforequestemails = self.nested_inforequestemail_queryset(obj)
+            outbound |= inforequestemails[0]
+            inbound |= inforequestemails[1]
+            if not obj_constraints:
+                if not obj.is_last_action and obj.next_action not in queryset:
+                    warnings.append(format_html(squeeze(u"""
+                            {} is not the last action in the branch. Deleting it may cause logical
+                            errors in the inforequest history.
+                            """).format(admin_obj_format(obj))))
+
+        if request.POST.get(u'post'):
+            if constraints:
+                raise PermissionDenied
+
+        template_response = delete_selected(self, request, queryset)
+
+        if request.POST.get(u'post'):
+            n = outbound.count()
+            if n:
+                outbound.delete()
+                self.message_user(request,
+                        u'Successfully deleted {} applicant_action inforequestemails.'.format(n),
+                        messages.SUCCESS)
+            m = inbound.count()
+            if m:
+                inbound.update(type=InforequestEmail.TYPES.UNDECIDED)
+                self.message_user(request,
+                        u'Successfully updated {} obligee_action inforequestemails.'.format(m),
+                        messages.SUCCESS)
+            return None
+
+        template_response.context_data.update({
+                u'outbound': [admin_obj_format(inforequestemail) for inforequestemail in outbound],
+                u'inbound': [admin_obj_format(inforequestemail) for inforequestemail in inbound],
+                u'delete_constraints': constraints,
+                u'warnings': warnings,
+        })
+        return template_response
```

Comment:
Toto by sa zislo spravit nejak vseobecnejsie, nech sa da pouzivat objektova composition / inheritance / delegation. Idealne asi ako mixin. Nieco ako `DeleteSelectedAdminMixin`.

Mixin bude mat hlavnu metodu pre samotnu akciu, ktora bude uvedena v `actions` atribute. Teda view metodu, analogicky ako v `ModelAdmin` je metoda `delete_view`. A potom tato metoda bude volat pomocne metody, ktore budu pri jednotlivych pouzitiach mixinu overridovane. Aby sa logika deletovania dala upravovat podla potreby. Analogicky ako to vieme pri standardnom deletovani.

Napriklad by sa zisla metoda, ktora doplni `context_data` do `template_response`. Aby sa dala overridnut analoficky ako metoda `render_delete_form` pri standardnom deletovani. 

Podobne by sa z hnevnej (view) metody zislo vydelit metodu na samotne mazanie z DB. Analogicky ako mame `delete_model` pri standardnom deletovani. Aby sme tuto metodu mohli overridovat a samotne mazanie upravovat podla potreby. Analogicky ako to robime s `delete_model`.

Potom by sme mohli rozsirit `DeleteNestedInforequestEmailAdminMixin` o overridy tychto novych metod, aby logika okolo mazania zavyslych inforequest emailov bola cela zapuzdrena do `DeleteNestedInforequestEmailAdminMixin`. Aby sme ju nemuseli implementovat na viacerych miestach. Ale aby stacilo do admin classy pridat `DeleteNestedInforequestEmailAdminMixin` a mazanie zavyslych inforequest emailov fungovalo out of box.

Reply:
@martinmacko47 Mixini a aj spolocne templaty by som napisal az po zapracovani admin mazania (jednotlivo aj hromadne) modelov `Branch`, `Inforequest`, `InforequestEmail`, `InforequestDraft`. Tieto modely budu mat tiez nieco spolocne a aj nejake specifika, a lahsie sa mi potom budu vytvarat mixiny pre vsetky tieto modeli naraz. Alebo predsa mam mixin vytvorit uz v tomto PR?

Reply:
:ok:  Mixiny teda sprav po dokonceni tychto styroch modelov.